### PR TITLE
Add login routing scenario code

### PR DESCRIPTION
**Pull request by johnpryan**
login-routing scenario

@jackkim9

### CODE REVIEW

File: nav2-usability/scenario-code/login-routing/router.dart [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+}
+
+class BookRouteInformationParser extends RouteInformationParser<AppRoutePath> {
+  @override
+  Future<AppRoutePath> parseRouteInformation(
+      RouteInformation routeInformation) async {
+    final uri = Uri.parse(routeInformation.location!);
+
+    // Handle '/'
+    if (uri.pathSegments.isEmpty) {
```
Comment (chunhtai) [inline nav2-usability/scenario-code/login-routing/router.dart:60]: We should do login authentication here. it is usually async request, we should have a mocked class something like this (I am not an expert on doing the auth flow)
```dart
class authentication {
  Future<bool> isSignedIn(dynamic myCookie); // Return true if the cookie contains previous login session.
  Future<String> userSignedIn(String userId, String password); // Returns a token represents the login session to be stored in the cookie
}
```

File: nav2-usability/scenario-code/login-routing/router.dart [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+
+class BooksApp extends StatefulWidget {
+  @override
+  State<StatefulWidget> createState() => _BooksAppState();
+}
+
+class AppState extends ChangeNotifier {
+  bool _isLoggedIn = false;
+
+  bool get isLoggedIn => _isLoggedIn;
```
Comment (chunhtai) [inline nav2-usability/scenario-code/login-routing/router.dart:15]: The login flow should not be in the app state. It should a wrapper over a (fake) http client.

Depends on how complex/realistic we want this to be, we may store some session token in the appstate.

Comment (johnpryan) [reply nav2-usability/scenario-code/login-routing/router.dart]: fixed - I added an Authentication class implemented by a MockAuthentication class

File: nav2-usability/scenario-code/sign-in-routing/router.dart [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final isSignedIn = _appState.auth.isSignedIn();
+    final viewingBooksScreen = _appState.viewingBooks;
+    return Navigator(
+      key: navigatorKey,
+      pages: [
+        if (isSignedIn) ...[
```
Comment (chunhtai) [inline nav2-usability/scenario-code/sign-in-routing/router.dart:157]: sign checking isSignedIn is usually an async operation in web application, we should do it during the route parsing

File: nav2-usability/scenario-code/sign-in-routing/router.dart [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+      title: 'Books App',
+      routerDelegate: _routerDelegate,
+      routeInformationParser: _routeInformationParser,
+    );
+  }
+}
+
+class BookRouteInformationParser extends RouteInformationParser<AppRoutePath> {
+  @override
+  Future<AppRoutePath> parseRouteInformation(
```
Comment (chunhtai) [inline nav2-usability/scenario-code/sign-in-routing/router.dart:90]: This method should check whether users isSignedIn and return SignInRoutePath if they are not

File: nav2-usability/scenario-code/sign-in-routing/router.dart [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+  final String username;
+  final String password;
+
+  Credentials(this.username, this.password);
+}
+
+abstract class Authentication {
+  bool isSignedIn();
+
+  void signOut();
```
Comment (chunhtai) [inline nav2-usability/scenario-code/sign-in-routing/router.dart:17]: Sign out is usually an async operation as well, but it is less important in this use case

File: nav2-usability/scenario-code/sign-in-routing/router.dart [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+
+class Credentials {
+  final String username;
+  final String password;
+
+  Credentials(this.username, this.password);
+}
+
+abstract class Authentication {
+  bool isSignedIn();
```
[the remark this answers is not in the document]
Comment (johnpryan) [reply nav2-usability/scenario-code/sign-in-routing/router.dart:15]: How would you return the correct value for `currentConfiguration` if checking that the user is signed in is async?

File: nav2-usability/scenario-code/sign-in-routing/router.dart [hunk 7]
```diff
@@ -0,0 +1,309 @@
...
+  @override
+  State<StatefulWidget> createState() => _BooksAppState();
+}
+
+class AppState extends ChangeNotifier {
+  final Authentication auth;
+
+  AppState(this.auth);
+
+  Future<bool> signIn(String username, String password) async {
```
[the remark this answers is not in the document]
Comment (johnpryan) [reply nav2-usability/scenario-code/sign-in-routing/router.dart:51]: to notify listeners I think

File: nav2-usability/scenario-code/sign-in-routing/router.dart (hunk 3 above)
Comment (johnpryan) [reply nav2-usability/scenario-code/sign-in-routing/router.dart:157]: Let me see if this works - making `isSignedIn` async could cause some problems with `currentConfiguration`

File: nav2-usability/scenario-code/sign-in-routing/router.dart (hunk 4 above)
Comment (johnpryan) [reply nav2-usability/scenario-code/sign-in-routing/router.dart:90]: Ah, gotcha.

File: nav2-usability/scenario-code/sign-in-routing/router.dart (hunk 6 above)
[the remark this answers is not in the document]
Comment (johnpryan) [reply nav2-usability/scenario-code/sign-in-routing/router.dart:15]: Nevermind - saw your comment below

**Comment by johnpryan** [comment]
@chunhtai this is ready for another look - I was able to make the route parser aware of the AppState so it can check whether the user is authenticated. I still felt like I need to add a new field to AppState to take care of whether or not the user is viewing the login screen, so I added `isViewingSignIn`.

**Review by chunhtai — APPROVED** [review]
LGTM

File: nav2-usability/scenario-code/sign-in-routing/router.dart [hunk 8]
```diff
@@ -0,0 +1,360 @@
...
+
+abstract class Authentication {
+  Future<bool> isSignedIn();
+
+  Future<void> signOut();
+
+  Future<bool> signIn(String username, String password);
+}
+
+class MockAuthentication implements Authentication {
```
Comment (nguyenxndaidev) [inline nav2-usability/scenario-code/sign-in-routing/router.dart:22]: Why don't we implement the authentication service with ChangeNotifier?

Comment (johnpryan) [reply nav2-usability/scenario-code/sign-in-routing/router.dart:22]: It's meant to represent an API like [GoogleSignIn](https://pub.dev/documentation/google_sign_in/latest/google_sign_in/GoogleSignIn-class.html)

Comment (nguyenxndaidev) [reply nav2-usability/scenario-code/sign-in-routing/router.dart:22]: ah, I see you use AppState to notify also auth status. So don't need a separated ChangeNotifier.

Comment (johnpryan) [reply nav2-usability/scenario-code/sign-in-routing/router.dart:22]: Yeah, the AppState notifies its listeners on behalf of the Authentication object changes.

**Pull request merged**